### PR TITLE
fix: allow for recursive block-on calls

### LINKED ISSUE #798: Bug: task::block_on cannot run on an async context
I have the following code:
```rust
// I don't control these parts. They come from a library:
async fn run() {
    while let Some(data) = receiver.next().await {
        process_data(data);
    }
}

// This is the part that I get to control
fn process_data(data: Vec<u8>) {
    // I need to run an async task here.
    // So I do this:
    task::block_on(async {
        do_something_async_here().await;
    });
}
```

The `process_data` function crashes with the following error: 
```
thread 'async-std/runtime' panicked at 'cannot run an executor inside another executor', <::std::macros::panic macros>:2:4
```

I think it might have to do with the switch to `smol`, since I didn't face this issue earlier with 1.5. This happens only on 1.6. Please advice me on how to proceed.

### PR DESCRIPTION
**Pull request by dignifiedquire**
Depends on https://github.com/stjepang/smol/pull/147

Fixes #798,#795,#760

### CODE REVIEW

**Comment by rakshith-ravi** [comment]
Awesome! Now that stjepang/smol#147 is merged, can we update this and merge this as well?

**Comment by dignifiedquire** [comment]
> Awesome! Now that [stjepang/smol#147](https://github.com/stjepang/smol/pull/147) is merged, can we update this and merge this as well?

still need to wait for a new release

**Comment by dignifiedquire** [comment]
CI seems to be failing independently, unclear as to why atm

File: src/task/builder.rs [hunk 1]
```diff
@@ -8,6 +9,11 @@ use pin_project_lite::pin_project;
 use crate::io;
 use crate::task::{JoinHandle, Task, TaskLocalsWrapper};
 
+#[cfg(not(target_os = "unknown"))]
+thread_local! {
+    static IS_CURRENT_BLOCKING: Cell<usize> = Cell::new(0);
```
Comment (yoshuawuyts) [inline src/task/builder.rs:14]: Hmm, not sure I love this name. Something like `BLOCK_ON_ENABLED` could work better. Or perhaps it's just missing docs.

File: src/task/builder.rs [hunk 2]
```diff
@@ -151,7 +157,36 @@ impl Builder {
...
+                is_current_blocking.replace(1);
+
+                // The first call should use run.
+                unsafe {
+                    TaskLocalsWrapper::set_current(&wrapped.tag, || {
+                        let res = smol::run(wrapped);
+                        is_current_blocking.replace(is_current_blocking.get() - 1);
+                        res
+                    })
+                }
```
Comment (yoshuawuyts) [inline src/task/builder.rs:173]: This code is duplicated with the branch below. It seems the only difference is between:
```rust
is_current_blocking.replace(1);
```
and
```rust
is_current_blocking.replace(is_current_blocking.get() + 1);
```
We should probably only branch on that and share the `TaskLocalsWrapper` logic.

File: tests/block_on.rs [hunk 3]
```diff
@@ -16,3 +16,62 @@ fn panic() {
...
+#[cfg(feature = "unstable")]
+#[test]
+fn nested_block_on_local() {
+    let x = task::block_on(async {
+        let a =
+            task::block_on(async { task::block_on(async { async_std::future::ready(3).await }) });
+        let b = task::spawn_local(async {
+            task::block_on(async { async_std::future::ready(2).await })
+        })
+        .await;
```
Comment (yoshuawuyts) [inline tests/block_on.rs:29]: This is kind of hard to follow. Can we split this on multiple lines or make it a one-liner? Same with the other statements below.

Maybe importing `use async_std::future` would be enough: that way we can write `future::ready(1).await` which will likely be enough to make this a one-liner.

Comment (dignifiedquire) [reply tests/block_on.rs:29]: fixed

File: src/task/builder.rs (hunk 2 above)
Comment (dignifiedquire) [reply src/task/builder.rs]: fixec

File: src/task/builder.rs (hunk 1 above)
Comment (dignifiedquire) [reply src/task/builder.rs]: fixed

**Comment by rakshith-ravi** [comment]
Awesome! Thanks a ton! Can we create a new release with this?

**Comment by rakshith-ravi** [comment]
@dignifiedquire I tried pointing the `async-std` in my Cargo.toml to the master branch and am still getting the same error as #798. Any idea what could be causing this?

**Comment by dignifiedquire** [comment]
might be pulling in a different version of smol or async-std somewhere?

**Comment by rakshith-ravi** [comment]
I modified my Cargo.toml to this:
```
async-std = { git = "https://github.com/async-rs/async-std", branch = "master", features = ["attributes"] }
```

Smol seems to be updated to 0.1.11, according to the Cargo.lock file. Any help on how to debug this would be appreciated. I'd be happy to test this out before we release the update

**Comment by dignifiedquire** [comment]
@rakshith-ravi I think I have an idea what the issue is, could you test out this branch: https://github.com/async-rs/async-std/tree/fix/recursive-block-2

**Comment by rakshith-ravi** [comment]
Hey @dignifiedquire. Sorry I've been MIA, got busy in a lot of work. That branch shows me a 404. Is there any other way for me to test it?

**Comment by rakshith-ravi** [comment]
Okay, just noticed that it got merged into master. And yes, @dignifiedquire I tested it with the current master branch and it seems to work well!
